### PR TITLE
enh(api) handle hostCategories in AddHostTemplate endpoint

### PR DESCRIPTION
**Pull request by a-launois**
## Description

Handle host categories in AddHostTemplate endpoint

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [x] 23.10.x (master)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 1]
```diff
@@ -17658,3 +17658,8 @@ msgstr "Vous n'êtes pas autorisé à lister les configurations de notifications
 msgid "Notification #%d should have at least one user"
 msgstr "La notification #%d doit avoir au moins un utilisateur"
 
+msgid "Error while linking host categories to a host template"
+msgstr "Erreur lors de la liaison de catégories d'hôte à un modèle d'hôte"
+
+msgid "The %s does not exist with id(s) '%s'"
```
Comment (cg-tw) [inline centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:17664]:
```suggestion
msgid "The %s does not exist with ID(s) '%s'"
```

File: centreon/src/Core/HostTemplate/Application/Exception/HostTemplateException.php [hunk 2]
```diff
@@ -86,6 +96,24 @@ public static function idDoesNotExist(string $propertieName, int $propertieValue
...
+     * @param string $propertieName
+     * @param int[] $propertieValues
+     *
+     * @return self
+     */
+    public static function idsDoNotExist(string $propertieName, array $propertieValues): self
+    {
+        return new self(
+            sprintf(
+                _("The %s does not exist with id(s) '%s'"),
```
Comment (cg-tw) [inline centreon/src/Core/HostTemplate/Application/Exception/HostTemplateException.php:109]:
```suggestion
                _("The %s does not exist with ID(s) '%s'"),
```

File: centreon/doc/API/v23.10/Configuration/HostTemplate/Schema/AddHostTemplateResponse.yaml [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+      * `1` - STATUS_ENABLED
+      * `2` - STATUS_DEFAULT (inheritance of its parent's value. If there is no parent, the values used will be that of Centreon Engine)
+  notification_options:
+    type: integer
+    description: |
+      Define the states of the host for which notifications should be sent out.
+
+      The value is the sum of all the values of the selected options.
+
+      * `1` - DOWN
```
Comment (callapa) [inline centreon/doc/API/v23.10/Configuration/HostTemplate/Schema/AddHostTemplateResponse.yaml:106]: No NONE value ?

**Review by abuathier — APPROVED** [review]

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php [hunk 4]
```diff
@@ -238,105 +244,67 @@ private function assertIsValidCommand(?int $commandId, ?CommandType $commandType
         }
     }
 
-    private function createNewHostTemplate(AddHostTemplateRequest $request): NewHostTemplate
+    /**
+     * Assert category IDs are valid.
+     *
+     * @param int[] $categoryIds
```
Comment (callapa) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php:250]:
```suggestion
     * @param int[] $categoryIds
     *
     * @throws HostTemplateException
     * @throws \Throwable
```

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php [hunk 5]
```diff
@@ -238,105 +244,67 @@ private function assertIsValidCommand(?int $commandId, ?CommandType $commandType
...
+
+        return $this->writeHostTemplateRepository->add($newHostTemplate);
     }
 
-    private function createResponse(?HostTemplate $hostTemplate): AddHostTemplateResponse
+    /**
+     * @param AddHostTemplateRequest $dto
+     * @param int $hostTemplateId
+     *
+     * @throws \Throwable
```
Comment (callapa) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php:298]:
```suggestion
     * @throws HostTemplateException
     * @throws \Throwable
```

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php [hunk 6]
```diff
@@ -238,105 +244,67 @@ private function assertIsValidCommand(?int $commandId, ?CommandType $commandType
...
+
+        if ([] !== ($invalidIds = array_diff($categoryIds, $validCategoryIds))) {
+            throw HostTemplateException::idsDoNotExist('categories', $invalidIds);
+        }
+    }
+
+    /**
+     * @param AddHostTemplateRequest $request
+     *
+     * @throws \Throwable
```
Comment (callapa) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php:269]:
```suggestion
     * @throws AssertionFailedException
     * @throws HostTemplateException
     * @throws \Throwable
```

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php [hunk 7]
```diff
@@ -83,30 +87,32 @@ public function __invoke(AddHostTemplateRequest $request, AddHostTemplatePresent
...
-            $this->assertIsValidTimePeriod($request->notificationTimeperiodId, 'notificationTimeperiodId');
-            $this->assertIsValidCommand($request->checkCommandId, CommandType::Check, 'checkCommandId');
-            $this->assertIsValidCommand($request->eventHandlerCommandId, null, 'eventHandlerCommandId');
-            $this->assertIsValidIcon($request->iconId);
+            try {
+                $this->dataStorageEngine->startTransaction();
 
-            $newHostTemplate = $this->createNewHostTemplate($request);
+                $hostTemplateId = $this->handleHostTemplate($request);
+                $this->handleHostCategories($request, $hostTemplateId);
```
Comment (a-launois) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php:94]:
```suggestion
                $hostTemplateId = $this->createHostTemplate($request);
                $this->linkHostCategories($request, $hostTemplateId);
```

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php [hunk 8]
```diff
@@ -238,105 +244,67 @@ private function assertIsValidCommand(?int $commandId, ?CommandType $commandType
...
+    }
+
+    /**
+     * @param AddHostTemplateRequest $request
+     *
+     * @throws \Throwable
+     *
+     * @return int
+     */
+    private function handleHostTemplate(AddHostTemplateRequest $request): int
```
Comment (a-launois) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php:273]:
```suggestion
    private function createHostTemplate(AddHostTemplateRequest $request): int
```

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php [hunk 9]
```diff
@@ -83,30 +87,32 @@ public function __invoke(AddHostTemplateRequest $request, AddHostTemplatePresent
...
+            }
+            if ($this->user->isAdmin()) {
+                $hostCategories = $this->readHostCategoryRepository->findByHost($hostTemplateId);
+            } else {
+                $accessGroups = $this->readAccessGroupRepository->findByContact($this->user);
+                $hostCategories = $this->readHostCategoryRepository->findByHostAndAccessGroups($hostTemplateId, $accessGroups);
             }
 
-            $presenter->presentResponse($this->createResponse($hostTemplate));
+            $presenter->presentResponse(AddHostTemplateFactory::createResponse($hostTemplate, $hostCategories));
```
Comment (callapa) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php:115]: Good idea : Use composition to complete an answer 👍

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php [hunk 10]
```diff
@@ -238,105 +244,67 @@ private function assertIsValidCommand(?int $commandId, ?CommandType $commandType
...
     }
 
-    private function createResponse(?HostTemplate $hostTemplate): AddHostTemplateResponse
+    /**
+     * @param AddHostTemplateRequest $dto
+     * @param int $hostTemplateId
+     *
+     * @throws \Throwable
+     */
+    private function handleHostCategories(AddHostTemplateRequest $dto, int $hostTemplateId): void
```
Comment (a-launois) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php:300]:
```suggestion
    private function linkHostCategories(AddHostTemplateRequest $dto, int $hostTemplateId): void
```

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php [hunk 11]
```diff
@@ -83,30 +87,32 @@ public function __invoke(AddHostTemplateRequest $request, AddHostTemplatePresent
...
-                );
-
-                return;
+                throw HostTemplateException::errorWhileRetrievingObject();
+            }
+            if ($this->user->isAdmin()) {
+                $hostCategories = $this->readHostCategoryRepository->findByHost($hostTemplateId);
+            } else {
+                $accessGroups = $this->readAccessGroupRepository->findByContact($this->user);
+                $hostCategories = $this->readHostCategoryRepository->findByHostAndAccessGroups($hostTemplateId, $accessGroups);
```
Comment (callapa) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php:112]: This line is longer than 120 characters, isn't it?

File: centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php [hunk 12]
```diff
@@ -238,105 +244,73 @@ private function assertIsValidCommand(?int $commandId, ?CommandType $commandType
...
-            $request->note,
-            $request->actionUrl,
-            $request->iconId,
-            $request->iconAlternative,
-            $request->comment,
-            $request->isActivated,
-        );
+        $newHostTemplate = NewHostTemplateFactory::create($request, $inheritanceMode);
+
+        return $this->writeHostTemplateRepository->add($newHostTemplate);
```
Comment (callapa) [inline centreon/src/Core/HostTemplate/Application/UseCase/AddHostTemplate/AddHostTemplate.php:296]: Logging is missing

**Review by callapa — APPROVED** [review]

**Pull request merged**